wallet/wallet.go

```diff
@@ -3465,11 +3474,8 @@ func (w *Wallet) ListUnspent(ctx context.Context, minconf, maxconf int32, addres
 			// This will be unnecessary once transactions and outputs are
 			// grouped under the associated account in the db.
 			acctName := defaultAccountName
-			sc, addrs, _, err := txscript.ExtractPkScriptAddrs(
-				0, output.PkScript, w.chainParams, true) // Yes treasury
-			if err != nil {
-				continue
-			}
+			sc, addrs := stdscript.ExtractAddrs(0 /* TODO */, output.PkScript, w.chainParams)
+			// if len(addrs) == 0 { continue }
```

Comment:
this is fine, if you want to refactor this to remove the indent

Reply:
Ah, sorry I meant to remove this comment.  I'm not confident that's the same behavior because the old `err != nil` is (and was) not the same thing as `if len(addrs) == 0`.

Reply:
It looks like the comment should be removed to me as it would potentially change the behavior.  Removing the err check without adding this is correct as the error return was **always** nil with the old code here because `txscript.ExtractPkScriptAddrs` only returned an error in the case the script version was not 0, but it was always being called with 0.

The old call could _also_ return an empty address slice if the scirpt had an invalid pubkey or the like in it.  So, that behavior remains the same.

Reply:
Looking at the wallet code closer, I'm pretty sure it actually would change the behavior to add this here since, if there are no addresses, the if branch won't be taken which would leave `acctName := defaultAccountName` and then the check just after says `if accountName != "" && accountName != acctName { continue }`, which means it would **not** skip to the next loop iteration immediately with the current code whereas the addition of this commented check would.

I think the existing code (and hence the updated code too) is buggy here actually.  The script class could theoretically have been e.g. `txscript.PubKeyHashTy` with `len(addrs) == 0` if a nonstandard transaction made it in.  Then, the existing code would end up executing through the switch statement later (since as noted above, it would not hit the continue statement), be identified as `txscript.PubKeyHashTy` thereby setting `spendable = true` and producing a result as if it is a spendable unspent output when it really isn't.

Reply:
To ensure that a `ListUnspentResult` is still appended in such a case (it is now), even if it is non-standard, I'm considering this change instead so that the entry is at least accurate w.r.t. `spendable` if address decoding fails for a script with a standard form:

```diff
- Spendable:     spendable,
+ Spendable:     spendable && len(addrs) > 0,
```

Reply:
That sounds reasonable.  I'm not entirely sure that something that will never be spendable by the wallet should be in the output, but I can see a case for that too since it is technically an unspent output tracked by the wallet and I agree that it currently has (or at least intends to have) those semantics.

Reply:
> I will note however that while wallet currently doesn't allow it as far as I know, there is no reason why a wallet couldn't theoretically spend some type of non-standard output that has no valid addresses associated with it.

Maybe this spendable change should be reverted, will see what @jrick wants, but I'd expect wallet to detect such special non-standard scripts and flip `spendable = true` (requiring this `spendable && len(addrs) > 0` to go away).

Reply:
We do still want outputs we don't know how to spend to appear in the listunspent results.

Reply:
OK, so code as-is will keep them, just set spendable = false at this spot.  Can change if you'd prefer it to read differently.
e.g.

```go
			// If address decoding failed, the output is not spendable
			// regardless of detected script type.
			spendable = spendable && len(addrs) > 0
```

_before_ constructing the `ListUnspentResult`.